core/src/main/scala/spire/optional/Perm.scala

```diff
@@ -0,0 +1,79 @@
+package spire
+package optional
+
+import scala.collection.compat._
+import spire.scalacompat.SeqLike
+
+import cats.kernel.Eq
+import spire.algebra.{Action, Group}
+import spire.algebra.partial.PartialAction
+import spire.syntax.cfor._
+import spire.syntax.group._
+import spire.util._
+
+/**
+ * Represents a permutation encoded as a map from preimages to images, including
+ * only pairs that are moved by the permutation (so the identity is Map.empty).
+ */
+class Perm private(val mapping: Map[Int, Int]) extends (Int => Int) {
+  assert(mapping.forall(Function.tupled(_ != _)), "Perm contains a redundant mapping.")
+  require(mapping.values.toSet == mapping.keySet, "Image and preimage must be the same.")
+
+  override def apply(k: Int): Int = mapping.getOrElse(k, k)
+
+  override def toString: String = {
+    mapping.toSeq.sorted
+      .map { case (k, v) => s"$k -> $v"}
+      .mkString("Perm(", ", ", ")")
+  }
+
+  def image: Set[Int] = mapping.keySet
+  def invert(k: Int): Int = mapping.find(_._2 == k).fold(k)(_._1)
```

Comment:
When you write the doc, I'd recommend calling out that this `invert` method is inefficient, in the sense of using linear search on each call. Contrast with `PermGroup.inverse` that generates the efficient inverse perm.

Reply:
Yeah, we'd need a real BiMap implementation to make this efficient. I'll make sure to note the performance in the doc.

Reply:
Actually, on second thought, that might be surprising to people expecting similar performance from `actl` and `actr`. It might be worth caching the inverse mapping privately.

Reply:
I had a similar issue one time, and IIRC I made an internal inverse, but I made it lazy, so it'd only be instantiated in the event someone called the `inverse` method